python/ray/scripts/symmetric_run.py

```diff
@@ -0,0 +1,271 @@
+
+    ray_start_args, entrypoint_on_head = all_args[:separator], all_args[separator + 1 :]
+
+    ray_start_args = clean_and_validate_ray_start_args(ray_start_args)
+    min_nodes = 1 if wait_for_nnodes is None else wait_for_nnodes
+
+    if check_ray_already_started():
+        raise click.ClickException("Ray is already started on this node.")
+
+    # 1. Parse address and check if we are on the head node.
+    gcs_ip_port = ray._common.network_utils.parse_address(address)
+    if gcs_ip_port is None:
+        raise click.ClickException(f"Invalid address format: {address}")
+    gcs_ip, gcs_port = gcs_ip_port
+
+    try:
+        # AF_UNSPEC allows resolving both IPv4 and IPv6
+        addrinfo = socket.getaddrinfo(
+            gcs_ip, gcs_port, socket.AF_UNSPEC, socket.SOCK_STREAM
+        )
+        resolved_gcs_ip = addrinfo[0][4][0]
+    except socket.gaierror:
+        raise click.ClickException(f"Could not resolve hostname: {gcs_ip}")
+
+    my_ips = []
+    for iface, addrs in psutil.net_if_addrs().items():
+        for addr in addrs:
+            # Look for AF_INET (IPv4) or AF_INET6 (IPv6)
+            if addr.family in [
+                socket.AddressFamily.AF_INET,
+                socket.AddressFamily.AF_INET6,
+            ]:
+                my_ips.append(addr.address)
+
+    # Add localhost ips if we are running on a single node.
+    if min_nodes > 1:
+        # Ban localhost ips if we are not running on a single node
+        # to avoid starting N head nodes
+        my_ips = [ip for ip in my_ips if ip != "127.0.0.1" and ip != "::1"]
+
+    is_head = resolved_gcs_ip in my_ips
+
+    result = None
+    # 2. Start Ray and run commands.
+    try:
+        if is_head:
+            # On the head node, start Ray, run the command, then stop Ray.
+            click.echo("On head node. Starting Ray cluster head...")
+
+            # Build the ray start command with all parameters
+            ray_start_cmd = [
+                "ray",
+                "start",
+                "--head",
+                f"--node-ip-address={resolved_gcs_ip}",
+                f"--port={gcs_port}",
+                *ray_start_args,
+            ]
+
+            # Start Ray head. This runs in the background and hides output.
+            subprocess.run(ray_start_cmd, check=True, capture_output=True)
+            click.echo("Head node started.")
+            click.echo("=======================")
+            if min_nodes > 1 and not check_cluster_ready(min_nodes):
+                raise click.ClickException(
+                    "Timed out waiting for other nodes to start."
+                )
+
+            # Run the user command if provided.
+            if entrypoint_on_head:
+                click.echo(
+                    f"Running command on head node: {entrypoint_on_head}",
+                )
+                click.echo("=======================")
+                result = subprocess.run(entrypoint_on_head)
```

Comment:
will the output of entrypoint be redirected to the current process?

Reply:
Yeah, this redirects to current process.